core/client/src/main/java/alluxio/client/block/stream/NettyPacketWriter.java

```diff
@@ -0,0 +1,443 @@
+    /**
+     * Default constructor.
+     */
+    public PacketWriteHandler() {}
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws IOException {
+      Preconditions.checkState(acceptMessage(msg), "Incorrect response type.");
+      RPCProtoMessage response = (RPCProtoMessage) msg;
+      Protocol.Status status = ((Protocol.Response) response.getMessage()).getStatus();
+
+      if (!Status.isOk(status)) {
+        throw new IOException(String
+            .format("Failed to write block %d from %s with status %s.", mId, mAddress,
+                status.toString()));
+      }
+      mLock.lock();
+      try {
+        mDone = true;
+        mDoneOrFailed.signal();
+      } finally {
+        mLock.unlock();
+      }
+    }
+
+    @Override
+    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
+      LOG.error("Exception caught while reading response from netty channel {}.",
+          cause);
+      mLock.lock();
+      try {
+        mPacketWriteException = cause;
+        mBufferNotFullOrFailed.signal();
+        mDoneOrFailed.signal();
+        mBufferEmptyOrFailed.signal();
+      } finally {
+        mLock.unlock();
+      }
+
+      ctx.close();
+    }
+
+    @Override
+    public void channelUnregistered(ChannelHandlerContext ctx) {
+      mLock.lock();
+      try {
+        if (mPacketWriteException == null) {
+          mPacketWriteException = new IOException("Channel closed.");
+        }
+        mBufferNotFullOrFailed.signal();
+        mDoneOrFailed.signal();
+        mBufferEmptyOrFailed.signal();
+      } finally {
+        mLock.unlock();
+      }
+      ctx.fireChannelUnregistered();
+    }
+
+    /**
+     * @param msg the message received
+     * @return true if this message should be processed
+     */
+    private boolean acceptMessage(Object msg) {
+      if (msg instanceof RPCProtoMessage) {
+        MessageLite header = ((RPCProtoMessage) msg).getMessage();
+        return header instanceof Protocol.Response;
+      }
+      return false;
+    }
+  }
+
+  /**
+   * The netty channel future listener that is called when packet write is flushed.
+   */
+  private final class WriteListener implements ChannelFutureListener {
+    private final long mPosToWriteUncommitted;
+
+    public WriteListener(long pos) {
+      mPosToWriteUncommitted = pos;
+    }
+
+    @Override
+    public void operationComplete(ChannelFuture future) {
+      if (!future.isSuccess()) {
+        future.channel().close();
+      }
+      boolean shouldSendEOF = false;
+      mLock.lock();
+      try {
+        Preconditions.checkState(mPosToWriteUncommitted - mPosToWrite <= PACKET_SIZE,
+            "Some packet is not acked.");
+        Preconditions.checkState(mPosToWriteUncommitted <= mLength);
+        mPosToWrite = mPosToWriteUncommitted;
+
+        if (future.cause() != null) {
+          mPacketWriteException = future.cause();
+          mDoneOrFailed.signal();
```

Comment:
There is only one thread listen on these condition variables. So signal() is sufficient. Do you see a case where signalAll has to be used?  I can change it to signalAll if you prefer. i agree that is safer but it can also hide bugs.